### PR TITLE
server, tenant: Hot ranges structured logging for serverless clusters

### LINKED ISSUE #85415: telemetry: track hot ranges from a cluster
**Is your feature request related to a problem? Please describe.**
The KV Observability team has several data questions that they'd like to answer that we don't have insights into today:
* Which org or cluster has the hottest range?
* How hot is that range? What database, table, index or node, store, sql pod, etc is that range associated with?
* What is the 50th, 70th, 90th-percentile QPS for hot ranges?

**Describe the solution you'd like**
Brainstorm: we could track hourly logs (or some other cadence) that take a snapshot of the currently hottest ranges (top X ranges) for each cloud cluster and their associated QPS, db, table, index, store, node, sql pod, etc. We could perhaps include other summary/aggregate stats alongside each telemetry log event. Details pending in a separate 1-pager

**Additional context**
We should backport this to 22.1 and 21.2

Jira issue: CRDB-18238

### PR DESCRIPTION
**Pull request by koorosh**
This patch introduces periodic logging of hot ranges info (ie QPS and
relevant DB objects info). It is done for both dedicated and
serverless cluster types, and by default called every 4h.
Stats collection can be enabled/disabled, or changed interval
through cluster settings.

Resolves: #85415

Release note: None

### CODE REVIEW

**Review by dhartunian — COMMENTED** [review]
> Reviewed 3 of 11 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89511)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @koorosh)

___
*[`pkg/server/tenant.go` line 737 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NNbN_17EK078geJRa9N:-NNbN_185RNBWVcef6s0:b-utwbku) ([raw file](https://github.com/cockroachdb/cockroach/blob/397bbac18c75bb1df309dfa5c2f371e3501cc0e6/pkg/server/tenant.go#L737)):*
> ```Go
> 	s.sqlServer.isReady.Set(true)
> 
> 	structlogging.StartHotRangesLoggingScheduler(ctx, s.stopper, s.sqlServer.tenantConnect, *s.sqlServer.internalExecutor)
> ```

should this be inside the SQLServer? 
___
*[`pkg/server/structlogging/hot_ranges_log.go` line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NNbNkuG1we_880VPiHd:-NNbNkuG1we_880VPiHe:b-u7kn3f) ([raw file](https://github.com/cockroachdb/cockroach/blob/397bbac18c75bb1df309dfa5c2f371e3501cc0e6/pkg/server/structlogging/hot_ranges_log.go#L49)):*
> ```Go
> func (s *hotRangesDetailsLoggingScheduler) start(ctx context.Context, stopper *stop.Stopper) {
> 	_ = stopper.RunAsyncTask(ctx, "hot-ranges-struct-logging", func(ctx context.Context) {
> 		ticker := time.NewTicker(5 * time.Second)
> ```

Commit says it's running once per hour but this says 5 seconds.

Do we want this configurable via a cluster setting? Or provide the ability to disable? I think we already have a setting that determines whether telemetry is enabled. Perhaps if that's not on, we don't start the async task at all.
___
*[`pkg/server/structlogging/hot_ranges_log_test.go` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NNbOWaE40p0o5-hfnYD:-NNbOWaE40p0o5-hfnYE:b3myek1) ([raw file](https://github.com/cockroachdb/cockroach/blob/397bbac18c75bb1df309dfa5c2f371e3501cc0e6/pkg/server/structlogging/hot_ranges_log_test.go#L6)):*
> ```Go
> 
> func TestStartScheduler(t *testing.T) {
> 	t.Skip()
> ```

I assume this needs a test :)

**Review by koorosh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89511)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[`pkg/server/tenant.go` line 737 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NNbN_17EK078geJRa9N:-NNbTqPT-XGvbdvvZmnJ:bhbggz7) ([raw file](https://github.com/cockroachdb/cockroach/blob/397bbac18c75bb1df309dfa5c2f371e3501cc0e6/pkg/server/tenant.go#L737)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

should this be inside the SQLServer? 
</blockquote></details>

My assumption was that we need to  structured logs for tenant servers only and didn't find a way to distinguish it inside of SqlServer.
Currently `SqlServer` exposes `preStart` and `startServeSQL` functions that doesn't look correct to init logging.
Do you think it'll be more appropriate to add `postStart` func that accepts args to identify if it is tenant or not and then init logging if necessary? ie 
```
func postStart (..., tenant bool) error {
    if tenant {
      structlogging.StartHotRangesLoggingScheduler(...
    }
}
```
___
*[`pkg/server/structlogging/hot_ranges_log.go` line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NNbNkuG1we_880VPiHd:-NNbXiBG-ZXJgbYQLF_2:b9a11xt) ([raw file](https://github.com/cockroachdb/cockroach/blob/397bbac18c75bb1df309dfa5c2f371e3501cc0e6/pkg/server/structlogging/hot_ranges_log.go#L49)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Commit says it's running once per hour but this says 5 seconds.

Do we want this configurable via a cluster setting? Or provide the ability to disable? I think we already have a setting that determines whether telemetry is enabled. Perhaps if that's not on, we don't start the async task at all.
</blockquote></details>

good point it is still under [discussion](https://docs.google.com/document/d/10qaYjXYXTkhk1Y_Yyy7W_clXKw9yyWD9bEiXXfBrgng/edit#heading=h.mdramtj3z51x) and should be updated
___
*[`pkg/server/structlogging/hot_ranges_log_test.go` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NNbOWaE40p0o5-hfnYD:-NNbYMAj4GLjeEWJtu-B:b-8pqas4) ([raw file](https://github.com/cockroachdb/cockroach/blob/397bbac18c75bb1df309dfa5c2f371e3501cc0e6/pkg/server/structlogging/hot_ranges_log_test.go#L6)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I assume this needs a test :)
</blockquote></details>

Will do ))

**Review by koorosh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89511)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[`pkg/server/structlogging/hot_ranges_log_test.go` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NNbOWaE40p0o5-hfnYD:-NPvVlP73iAdaYIZed9b:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/397bbac18c75bb1df309dfa5c2f371e3501cc0e6/pkg/server/structlogging/hot_ranges_log_test.go#L6)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

Will do ))
</blockquote></details>

Done.

**Review by dhartunian — COMMENTED** [review]
> Reviewed 7 of 12 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89511)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @koorosh)

___
*[`pkg/server/tenant.go` line 737 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NNbN_17EK078geJRa9N:-NPwolI47B2_adgnDOCo:b3x2xmx) ([raw file](https://github.com/cockroachdb/cockroach/blob/397bbac18c75bb1df309dfa5c2f371e3501cc0e6/pkg/server/tenant.go#L737)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

My assumption was that we need to  structured logs for tenant servers only and didn't find a way to distinguish it inside of SqlServer.
Currently `SqlServer` exposes `preStart` and `startServeSQL` functions that doesn't look correct to init logging.
Do you think it'll be more appropriate to add `postStart` func that accepts args to identify if it is tenant or not and then init logging if necessary? ie 
```
func postStart (..., tenant bool) error {
    if tenant {
      structlogging.StartHotRangesLoggingScheduler(...
    }
}
```
</blockquote></details>

:thumbsup: no need to create a separate phase. we can leave it here, thx for the notes.
___
*[`pkg/server/structlogging/hot_ranges_log.go` line 11 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NPwotFd8znD_TKmky3b:-NPwotFd8znD_TKmky3c:b6u30f0) ([raw file](https://github.com/cockroachdb/cockroach/blob/404f4ef23788fc28646d7e199c38bb126603ea9f/pkg/server/structlogging/hot_ranges_log.go#L11)):*
> ```Go
> // licenses/APL.txt.
> 
> package structlogging
> ```

nit: can we use a different name? this seems too generic. maybe just `kvtelemetry`? or can we put this in an existing telemetry package somewhere?
___
*[`pkg/server/structlogging/hot_ranges_log.go` line 46 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NPwzbtzBYeBDhYt8Jjo:-NPwzbtzBYeBDhYt8Jjp:bqcxg8l) ([raw file](https://github.com/cockroachdb/cockroach/blob/404f4ef23788fc28646d7e199c38bb126603ea9f/pkg/server/structlogging/hot_ranges_log.go#L46)):*
> ```Go
> )
> 
> var TelemetryHotRangesStatsLoggingDelay = settings.RegisterDurationSetting(
> ```

do we need to respond to these cluster settings changing?
___
*[`pkg/server/structlogging/hot_ranges_log.go` line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NPwp4dUCVArW8XSI4U_:-NPwp4dUCVArW8XSI4Ua:b-lg1kbt) ([raw file](https://github.com/cockroachdb/cockroach/blob/404f4ef23788fc28646d7e199c38bb126603ea9f/pkg/server/structlogging/hot_ranges_log.go#L56)):*
> ```Go
> // hotRangesDetailsLoggingScheduler is responsible for logging index usage stats
> // on a scheduled interval.
> type hotRangesDetailsLoggingScheduler struct {
> ```

nit: can we rename to `hotRangesLoggingScheduler` to match the function? I don't think we need "Details".
___
*[`pkg/server/structlogging/hot_ranges_log.go` line 121 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NPx-KO7AwBF_p2iPwB-:-NPx-KO7AwBF_p2iPwB0:b-irz7e8) ([raw file](https://github.com/cockroachdb/cockroach/blob/404f4ef23788fc28646d7e199c38bb126603ea9f/pkg/server/structlogging/hot_ranges_log.go#L121)):*
> ```Go
> }
> 
> func logEventsWithDelay(
> ```

Can you add a comment about why we want to rate limit this? It's not immediately obvious to me. the events are already generated.

**Review by koorosh — COMMENTED** [review]
> Dismissed @dhartunian from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89511)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[`pkg/server/structlogging/hot_ranges_log.go` line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NNbNkuG1we_880VPiHd:-NPxiQeP7ayDZYp1z-Th:bm0smbn) ([raw file](https://github.com/cockroachdb/cockroach/blob/397bbac18c75bb1df309dfa5c2f371e3501cc0e6/pkg/server/structlogging/hot_ranges_log.go#L49)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

good point it is still under [discussion](https://docs.google.com/document/d/10qaYjXYXTkhk1Y_Yyy7W_clXKw9yyWD9bEiXXfBrgng/edit#heading=h.mdramtj3z51x) and should be updated
</blockquote></details>

FYI: Updated to once per day
___
*[`pkg/server/structlogging/hot_ranges_log.go` line 11 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NPwotFd8znD_TKmky3b:-NPxf8ya4uZkS1pGb_fC:b-rjqne6) ([raw file](https://github.com/cockroachdb/cockroach/blob/404f4ef23788fc28646d7e199c38bb126603ea9f/pkg/server/structlogging/hot_ranges_log.go#L11)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: can we use a different name? this seems too generic. maybe just `kvtelemetry`? or can we put this in an existing telemetry package somewhere?
</blockquote></details>

As I see, another packages define such logic in separate file as part of package where it is used (ie `pkg/ccl/changefeedccl/telemetry.go`).
In our case, I could lift it up on one level up in `server` package, and rename `hot_ranges_log.go` to `telemetry.go`. WDYT?
___
*[`pkg/server/structlogging/hot_ranges_log.go` line 46 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NPwzbtzBYeBDhYt8Jjo:-NPxhNj0ACpojbTxt4mn:b2qp2e9) ([raw file](https://github.com/cockroachdb/cockroach/blob/404f4ef23788fc28646d7e199c38bb126603ea9f/pkg/server/structlogging/hot_ranges_log.go#L46)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

do we need to respond to these cluster settings changing?
</blockquote></details>

I think it is not necessary. This cluster setting is fetched every time on new cycle of logging stats, so updated value will be used:
```
logEventsWithDelay(ctx, events, stopper, TelemetryHotRangesStatsLoggingDelay.Get(&s.st.SV))
```
___
*[`pkg/server/structlogging/hot_ranges_log.go` line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NPwp4dUCVArW8XSI4U_:-NPxiBzkB35-M8-FCBPk:bebrt1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/404f4ef23788fc28646d7e199c38bb126603ea9f/pkg/server/structlogging/hot_ranges_log.go#L56)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: can we rename to `hotRangesLoggingScheduler` to match the function? I don't think we need "Details".
</blockquote></details>

Will do
___
*[`pkg/server/structlogging/hot_ranges_log.go` line 121 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NPx-KO7AwBF_p2iPwB-:-NPxmOdi9JPuhfFphND0:bc1sgm5) ([raw file](https://github.com/cockroachdb/cockroach/blob/404f4ef23788fc28646d7e199c38bb126603ea9f/pkg/server/structlogging/hot_ranges_log.go#L121)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Can you add a comment about why we want to rate limit this? It's not immediately obvious to me. the events are already generated.
</blockquote></details>

Telemetry logging has limitation to send 10 log lines (with limited length) per second on each node and sending multiple entries (each can have arbitrary length) at the same time. This func allows defer sending logs with provided delay to make sure it doesn't exceed limits or blocks others to send logs.
I borrowed this func from `pkg/sql/scheduledlogging/captured_index_usage_stats.go` so going to move it to `logutil` package to have single version of this helper func.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89511)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @koorosh)

___
*[`pkg/server/structlogging/hot_ranges_log.go` line 121 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NPx-KO7AwBF_p2iPwB-:-NPyAaX0B5AZdWOD8fah:b-rujvx6) ([raw file](https://github.com/cockroachdb/cockroach/blob/404f4ef23788fc28646d7e199c38bb126603ea9f/pkg/server/structlogging/hot_ranges_log.go#L121)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

Telemetry logging has limitation to send 10 log lines (with limited length) per second on each node and sending multiple entries (each can have arbitrary length) at the same time. This func allows defer sending logs with provided delay to make sure it doesn't exceed limits or blocks others to send logs.
I borrowed this func from `pkg/sql/scheduledlogging/captured_index_usage_stats.go` so going to move it to `logutil` package to have single version of this helper func.
</blockquote></details>

Thanks. My recommendation however would be to pre-populate the event.Timestamp field (instead of relying on log.StructuredEvent doing it), so that the timestamp represents the time at which the HotRanges request was performed and not the time at which the event was logged.

**Review by koorosh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89511)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @knz)

___
*[`pkg/server/structlogging/hot_ranges_log.go` line 121 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NPx-KO7AwBF_p2iPwB-:-NQ-l-Xp5MGBvEq2rs4Z:be7qhbd) ([raw file](https://github.com/cockroachdb/cockroach/blob/404f4ef23788fc28646d7e199c38bb126603ea9f/pkg/server/structlogging/hot_ranges_log.go#L121)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Thanks. My recommendation however would be to pre-populate the event.Timestamp field (instead of relying on log.StructuredEvent doing it), so that the timestamp represents the time at which the HotRanges request was performed and not the time at which the event was logged.
</blockquote></details>

Thank you! Done.

**Comment by koorosh** [comment]
@thtruo , I set following default values for cluster settings that manage how logs are collected:
- `server.telemetry.hot_ranges_stats.interval` = 24h (collect stats once per day)
- `server.telemetry.hot_ranges_stats.enabled` is `True` by default (allows to disable/enable hot ranges stats collection)

also it's guarded by `diagnostics.reporting.enabled` setting which as I understand is general approach to enable reporting.

Let me know if it looks ok or should be changed.

**Review by kvoli — COMMENTED** [review]
I had a few questions about this that maybe @koorosh  or @thtruo could help me with:

The PR and commit message mention "...for serverless clusters" and "...for multi-tenant clusters".  Would it be possible to add this generally? Or do we already get this under some unified architecture stuff where every cluster is at least single tenant?

The work here is great, It would be nice have more frequent logging for debugging purposes (think debug.zip). Do you have any thoughts on https://github.com/cockroachdb/cockroach/issues/98108? It feels as though there is overlap and that issue is more ad-hoc. The period required for that issue is much lower and doesn't relate to telemetry. There seems like potential from this PR to perform more frequent (non-telemetry) logging for a subset of ranges into a log file similar to the [health log](https://github.com/cockroachdb/cockroach/blob/deb3f94e90930f998cce1687a2d17fcdb41787fb/pkg/server/status/runtime_log.go#L47-L47). 

Why is the log period 24h? I assume for telemetry you wish to capture seasonality. A common form of seasonality I've seen on user clusters takes place <24h e.g. people buying items at night and very little in the morning. It would be a shame for the telemetry data a cluster emits to be specific to the start time of the SQL tenant cluster.
   - I suggest jittering or randomizing when logging occurs, with a budget for how frequent logging can be in a period.
   - I suggest lowering the period from 24h to 4h or less.

My questions and suggestions shouldn't block this PR.

> Reviewed 9 of 12 files at r2, 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-:-NQ6smb305O6vIEFO5Kp:b-qwistt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @knz, and @koorosh)

**Review by koorosh — COMMENTED** [review]
Thanks for really great questions to discuss!

> commit message mention "...for serverless clusters" and "...for multi-tenant clusters"

I haven't considered to generalize it for UA and agree it is the right way to do but would still prefer to file separate issue for this.


Regarding interval durations, there is more question to Tommy and how this stats will be used at the end. I imagine that it shouldn't be called often because:
a) for seasonal increase of requests - it is more appropriate to think about the ways we can export hot ranges stats to prometheus at least so **use**r can monitor this data and react on such spikes, because we don't even know if this spice of requests expected or not.
b) telemetry might make use of infrequent stats collection to observe if there's some gradual regression (ie increased QPS) or to correlate hot ranges stats to schema change stats for instance. 

@thtruo what's your thoughts on this? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-:-NQ9whr-6uAFXa-R1GOl:b-70vgyg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @knz, and @kvoli)

**Comment by thtruo** [comment]
**Re time intervals** - I’m okay collecting more often in a tighter time interval. I’m not sure how randomizing/jittering will look, but I’d be okay starting with a fixed interval that’s more frequent. I was originally open to collecting every 8hrs, but based on Austen’s comment I’m also okay with a 4hr interval (as long as we don’t hit the logging pipeline’s constraints) - so let’s do that too.

@koorosh Will this telemetry fire for Dedicated clusters? The above comments suggest this telemetry will be applied to Serverless/multi-tenant clusters. Can you confirm? I ideally want to see this telemetry for **all** cloud clusters (including Dedicated and Serverless) - the original issue [#85415](https://github.com/cockroachdb/cockroach/issues/85415) does not specify only Serverless. Would capturing Dedicated telemetry come in a separate PR?

**Comment by koorosh** [comment]
Updated PR to collect telemetry for all cluster types and changed default interval to 4 hours.

**Review by kvoli — COMMENTED** [review]
This LGTM however I'm not that familiar with best practices in this area of the code. I'll defer to @dhartunian or @knz  for stamping.

I would suggest updating the PR title to mention the functionality is for all clusters.

> Reviewed 6 of 6 files at r4, 5 of 5 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-:-NRZdf-D6qFEdFm-Jd0o:bgzytm6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @koorosh)

___
*[`-- commits` line 2 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NRZceaB9Zm_ZGm8lbQQ):*
Are you planning on squashing the two commits together?
___
*[`-- commits` line 4 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NRZbc2xAyIpYTAfeCoS):*
nit: reword to `This patch introduces periodic logging of hot ranges info...`

**Review by kvoli — APPROVED** [review]
I'm going to stamp this, as it LGTM and but perhaps reach out to the other reviewers.

Are you going to backport this to 23.1?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-:-NRiL1eq-8ff79umorue:b-ppr73r)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @knz, and @koorosh)

**Comment by koorosh** [comment]
Yes, it has to be backported, almost forgot about this!

I'll wait for another pair of eyes on this, TFTR!

**Review by dhartunian — COMMENTED** [review]
Just have some minor comments on code organization and stuff and then it should be good to merge.

> Reviewed 1 of 12 files at r2, 3 of 6 files at r3, 5 of 6 files at r4, 3 of 5 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-:-NTRm-RR-Zg5E3h7mi_f:b-ofddsb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @koorosh)

___
*[`pkg/server/server_sql.go` line 1677 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NTP20FtAGEf1rQQMeMp:-NTP20FtAGEf1rQQMeMq:bf6nsaq) ([raw file](https://github.com/cockroachdb/cockroach/blob/45cf6b51b74d594bf72fd34158d34474a656d7d1/pkg/server/server_sql.go#L1677)):*
> ```Go
> 	pgL net.Listener,
> 	socketFileCfg *string,
> 	sServer serverpb.TenantStatusServer,
> ```

nit: call this `statusServer` for clarity
___
*[`pkg/server/tenant.go` line 737 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NNbN_17EK078geJRa9N:-NTRhgR-20w0bdpe69kw:b-d6v1dd) ([raw file](https://github.com/cockroachdb/cockroach/blob/397bbac18c75bb1df309dfa5c2f371e3501cc0e6/pkg/server/tenant.go#L737)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

:thumbsup: no need to create a separate phase. we can leave it here, thx for the notes.
</blockquote></details>

I think I preferred the placement of the scheduler here instead of `startServeSQL` just to keep the SQL startup as minimal as possible. It's fine if the code is duplicated in `server.go`.

For instance, you end up passing in `sqlServer.internalExecutor` to `startServeSQL`, which is a bit confusing.
___
*[`pkg/server/structlogging/hot_ranges_log.go` line 46 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NPwzbtzBYeBDhYt8Jjo:-NTRi7X541knmvgJOz8p:bw5hc66) ([raw file](https://github.com/cockroachdb/cockroach/blob/404f4ef23788fc28646d7e199c38bb126603ea9f/pkg/server/structlogging/hot_ranges_log.go#L46)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

I think it is not necessary. This cluster setting is fetched every time on new cycle of logging stats, so updated value will be used:
```
logEventsWithDelay(ctx, events, stopper, TelemetryHotRangesStatsLoggingDelay.Get(&s.st.SV))
```
</blockquote></details>

Great. Thanks!
___
*[`pkg/server/structlogging/hot_ranges_log.go` line 91 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NTP4i-87rcZoX4uyFFR:-NTP4i-87rcZoX4uyFFS:b-cczfi3) ([raw file](https://github.com/cockroachdb/cockroach/blob/45cf6b51b74d594bf72fd34158d34474a656d7d1/pkg/server/structlogging/hot_ranges_log.go#L91)):*
> ```Go
> 			select {
> 			case <-stopper.ShouldQuiesce():
> 				return
> ```

nit: also respond to `ctx` cancellation

**Review by koorosh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89511)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @knz, and @kvoli)

___
*[`-- commits` line 2 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NRZceaB9Zm_ZGm8lbQQ):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Are you planning on squashing the two commits together?
</blockquote></details>

Done.
___
*[`-- commits` line 4 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NRZbc2xAyIpYTAfeCoS):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

nit: reword to `This patch introduces periodic logging of hot ranges info...`
</blockquote></details>

Done.
___
*[`pkg/server/tenant.go` line 737 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NNbN_17EK078geJRa9N:-NTSkqN403UFq4Vyl3RF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/397bbac18c75bb1df309dfa5c2f371e3501cc0e6/pkg/server/tenant.go#L737)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I think I preferred the placement of the scheduler here instead of `startServeSQL` just to keep the SQL startup as minimal as possible. It's fine if the code is duplicated in `server.go`.

For instance, you end up passing in `sqlServer.internalExecutor` to `startServeSQL`, which is a bit confusing.
</blockquote></details>

Done.
___
*[`pkg/server/structlogging/hot_ranges_log.go` line 91 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NTP4i-87rcZoX4uyFFR:-NTSl7UA8YIQq4vFcBgv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/45cf6b51b74d594bf72fd34158d34474a656d7d1/pkg/server/structlogging/hot_ranges_log.go#L91)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: also respond to `ctx` cancellation
</blockquote></details>

Done.
___
*[`pkg/server/server_sql.go` line 1677 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NTP20FtAGEf1rQQMeMp:-NTSi_1v1GNsCH90s4Dp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/45cf6b51b74d594bf72fd34158d34474a656d7d1/pkg/server/server_sql.go#L1677)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: call this `statusServer` for clarity
</blockquote></details>

Done.

**Review by dhartunian — APPROVED** [review]
:lgtm: thanks @koorosh !

> Reviewed 2 of 8 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-:-NTTs0hn5yKZ8xqkA4Zq:btjr7k6)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz, @koorosh, and @kvoli)

___
*[`pkg/server/structlogging/hot_ranges_log.go` line 11 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/89511#-NPwotFd8znD_TKmky3b:-NTTqEIL7CTP31ptM3j3:b-cswduk) ([raw file](https://github.com/cockroachdb/cockroach/blob/404f4ef23788fc28646d7e199c38bb126603ea9f/pkg/server/structlogging/hot_ranges_log.go#L11)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

As I see, another packages define such logic in separate file as part of package where it is used (ie `pkg/ccl/changefeedccl/telemetry.go`).
In our case, I could lift it up on one level up in `server` package, and rename `hot_ranges_log.go` to `telemetry.go`. WDYT?
</blockquote></details>

Let's leave this as-is.

**Pull request merged**
